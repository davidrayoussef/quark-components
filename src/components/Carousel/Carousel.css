.wrapper {
  display: grid;
  grid-template-columns: 100px 1fr 100px;
  max-width: 730px;
  justify-items: center;
  align-items: center;
}

.images {
  position: relative;
  overflow: hidden;
}

.image-container {
  display: flex;
  align-items: flex-start;
}

.wrapper svg {
  transition: fill .15s;
}

.wrapper svg:hover {
  fill: rgba(0, 0, 0, .4);
}

.wrapper img {
  width: 100%;
  min-width: 100%;
  object-fit: contain;
}
