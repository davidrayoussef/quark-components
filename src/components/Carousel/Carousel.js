import React, { Component } from 'react';
import PropTypes from 'prop-types';
import Icon from '../Icon/Icon';
import style from './Carousel.css';

class Carousel extends Component {
  state = {
    activeIndex: 0
  };

  static defaultProps = {
    imgWidth: 900
  };

  static propTypes = {
    images: PropTypes.arrayOf(PropTypes.object).isRequired,
    imgWidth: PropTypes.number
  };

  handleLeftArrowClick = (e) => {
    console.log(e.target);
  };

  handleRightArrowClick = (e) => {
    console.log(e.target);
  };

  render() {
    const { images, imgWidth } = this.props;

    const renderedImages = images.map(img =>
      <img key={img.title} src={img.src} alt={img.title} />
    );

    if (renderedImages.length) {
      return (
        <main className={style.wrapper}>
          <Icon value="arrowLeft" onClick={this.handleLeftArrowClick} />

          <div className={style['image-container']}>
            { renderedImages }
          </div>

          <Icon value="arrowRight" onClick={this.handleRightArrowClick} />
        </main>
      );
    }
    return null;
  }
}

export default Carousel;
