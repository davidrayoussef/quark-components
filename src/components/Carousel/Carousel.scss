.carousel {
  display: grid;
  grid-template-columns: 50px 1fr 50px;
  max-width: 730px;
  justify-items: center;
  align-items: center;
  & .outer-container {
    position: relative;
    width: 100%;
    overflow: hidden;
    & .inner-container {
      white-space: nowrap;
    }
  }
  & .arrow-icon {
    cursor: pointer;
    width: 30px;
    transition: fill 0.15s;
    &:hover {
      fill: rgba(0, 0, 0, 0.4);
    }
  }
}
