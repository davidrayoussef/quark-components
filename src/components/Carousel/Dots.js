import React from 'react';
import PropTypes from 'prop-types';
import style from './Dots.css';

const Dots = ({ images, activeIndex, containerWidth, slideToIndex }) => (
  <div className={style['dots-container']} style={{ width: containerWidth }}>
    {
      images.map((item, i) => (
        <span
          key={item.title}
          className={ activeIndex === i ? style.active : '' }
          onClick={slideToIndex(i)}
          style={{ width: containerWidth / 35, height: containerWidth / 35 }}
        >
        </span>
      ))
    }
  </div>
);

Dots.propTypes = {
  images: PropTypes.array.isRequired,
  activeIndex: PropTypes.number.isRequired,
  containerWidth: PropTypes.number.isRequired,
  slideToIndex: PropTypes.func.isRequired
};

export default Dots;
