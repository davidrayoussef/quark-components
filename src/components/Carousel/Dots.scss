.dots-container {
  --primary: rgba(255, 255, 255, .6);
}

.dots-container {
  position: absolute;
  bottom: 20px;
  display: flex;
  align-items: flex-end;
  justify-content: center;
  color: var(--primary);
}

.dots-container span {
  margin: 0 6px;
  border-radius: 100%;
  border: 1px solid var(--primary);
  cursor: pointer;
}

.active {
  background: #FFF;
}
