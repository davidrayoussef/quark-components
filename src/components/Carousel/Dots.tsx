import React from 'react';

import { DotsProps } from '@/shared';

import style from './Dots.scss';

export const Dots: React.VFC<DotsProps> = ({
  activeIndex,
  childrenCount,
  handleNavClick
}: DotsProps) => (
  <div className={style.dotsContainer}>
    {Array.from({ length: childrenCount }, (_: void, index: number) => (
      <span
        key={index}
        className={activeIndex === index ? style.active : ''}
        onClick={handleNavClick.bind(undefined, index)}
      ></span>
    ))}
  </div>
);
