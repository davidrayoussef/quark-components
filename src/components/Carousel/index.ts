export { Carousel } from './Carousel';
export { Dots } from './Dots';
