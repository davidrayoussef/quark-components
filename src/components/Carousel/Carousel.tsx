import React, { Component, Fragment } from 'react';

import { Dots } from './Dots';
import { Icon } from '../Icon';
import { Image } from '../Image';
import { Swiper } from '../Swiper';

import { CarouselProps, CarouselState, SwiperProps } from '../../shared';

import style from './Carousel.css';

export class Carousel extends Component<CarouselProps, CarouselState> {
  imageContainerElement: React.RefObject<HTMLDivElement> = React.createRef();

  state: Readonly<CarouselState> = {
    activeIndex: 0,
    translate: 0,
    imgWidth: 0,
    shouldAnimate: true
  };

  static defaultProps = {
    showArrows: true,
    showDots: false,
    useSwiper: true
  };

  componentDidMount(): void {
    window.addEventListener('resize', this.handleResize);
  }

  componentWillUnmount(): void {
    window.removeEventListener('resize', this.handleResize);
  }

  setImageWidth = (): void => {
    this.setState({
      imgWidth: this.imageContainerElement.current.offsetWidth
    });
  };

  handleResize = (): void => {
    const { activeIndex, translate } = this.state;
    this.setState({
      imgWidth: this.imageContainerElement.current.offsetWidth,
      translate:
        Math.sign(translate) *
        (activeIndex * this.imageContainerElement.current.offsetWidth),
      shouldAnimate: false
    });
  };

  slidePrev = (): void => {
    const { activeIndex, translate } = this.state;
    const imgCount = this.props.images.length;
    const imgWidth =
      this.imageContainerElement.current &&
      this.imageContainerElement.current.offsetWidth;
    this.setState({
      translate:
        activeIndex === 0
          ? translate - imgWidth * (imgCount - 1)
          : translate + imgWidth,
      activeIndex: activeIndex === 0 ? imgCount - 1 : activeIndex - 1,
      shouldAnimate: true
    });
  };

  slideNext = (): void => {
    const { activeIndex, translate } = this.state;
    const imgCount = this.props.images.length;
    const imgWidth =
      this.imageContainerElement.current &&
      this.imageContainerElement.current.offsetWidth;
    this.setState({
      translate: (translate - imgWidth) % (imgWidth * imgCount),
      activeIndex: (activeIndex + 1) % imgCount,
      shouldAnimate: true
    });
  };

  slideToIndex = (index: number) => (): void => {
    const { activeIndex, translate, imgWidth } = this.state;
    this.setState({
      translate:
        translate +
        (index > activeIndex
          ? -((index - activeIndex) * imgWidth)
          : (activeIndex - index) * imgWidth),
      activeIndex: index,
      shouldAnimate: true
    });
  };

  render(): JSX.Element {
    const { activeIndex, translate, imgWidth, shouldAnimate } = this.state;
    const { images, showArrows, showDots, useSwiper } = this.props;
    const renderedImages = images.map(({ title, src }) => (
      <Image key={title} src={src} alt={title} onLoad={this.setImageWidth} />
    ));
    const arrowIconStyle = {
      cursor: 'pointer',
      minWidth: 25,
      maxWidth: 50,
      width: '7vw'
    };
    let Component;
    const props: Pick<SwiperProps, 'onSwipeLeft' | 'onSwipeRight'> = {};
    if (useSwiper) {
      Component = Swiper;
      props.onSwipeLeft = this.slideNext;
      props.onSwipeRight = this.slidePrev;
    } else Component = Fragment;
    return renderedImages.length ? (
      <section
        className={style.wrapper}
        style={!showArrows ? { gridTemplateColumns: '1fr' } : {}}
      >
        {showArrows && (
          <Icon
            value="arrowLeft"
            color="lightgray"
            style={arrowIconStyle}
            onClick={this.slidePrev}
          />
        )}
        <div className={style.images}>
          <Component {...props}>
            <div
              ref={this.imageContainerElement}
              className={style.imageContainer}
              style={{
                transform: `translateX(${translate}px)`,
                transition: shouldAnimate ? 'transform .4s ease-in-out' : 'none'
              }}
            >
              {renderedImages}
            </div>
          </Component>
          {showDots && (
            <Dots
              images={images}
              activeIndex={activeIndex}
              containerWidth={imgWidth}
              slideToIndex={this.slideToIndex}
            />
          )}
        </div>
        {showArrows && (
          <Icon
            value="arrowRight"
            color="lightgray"
            style={arrowIconStyle}
            onClick={this.slideNext}
          />
        )}
      </section>
    ) : null;
  }
}
