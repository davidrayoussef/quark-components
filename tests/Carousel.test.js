import React from 'react';
import { shallow, mount } from 'enzyme';
import Carousel from '../src/components/Carousel';

const images = [
  { title: 'image 1', src: 'images/slide-1.jpg' },
  { title: 'image 2', src: 'images/slide-2.jpg' },
  { title: 'image 3', src: 'images/slide-3.jpg' }
];

test('should render', () => {
  const wrapper = shallow(
    <Carousel images={[]} />
  );
  expect(wrapper);
});

test('should render 3 images', () => {
  const wrapper = mount(
    <Carousel images={images} />
  );
  expect(wrapper.find('img').length).toBe(3);
});

test('should correctly update activeIndex on left arrow click', () => {
  const wrapper = shallow(
    <Carousel images={images} />
  );
  const leftButton = wrapper.children().at(0);
  leftButton.simulate('click');
  expect(wrapper.state().activeIndex).toBe(2);
});

test('should correctly update activeIndex on right arrow click', () => {
  const wrapper = shallow(
    <Carousel images={images} />
  );
  const rightButton = wrapper.children().at(2);
  rightButton.simulate('click');
  expect(wrapper.state().activeIndex).toBe(1);
});

test('snapshot', () => {
  const wrapper = shallow(
    <Carousel images={images} />
  );
  expect(wrapper).toMatchSnapshot();
});
