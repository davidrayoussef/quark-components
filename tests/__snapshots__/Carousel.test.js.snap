// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot 1`] = `
<section
  className="wrapper"
  style={Object {}}
>
  <Icon
    color="lightgray"
    onClick={[Function]}
    style={
      Object {
        "cursor": "pointer",
        "maxWidth": 50,
        "minWidth": 25,
        "width": "7vw",
      }
    }
    value="arrowLeft"
    width={60}
  />
  <div
    className="images"
  >
    <Swiper
      onSwipeLeft={[Function]}
      onSwipeRight={[Function]}
      threshold={15}
    >
      <div
        className="image-container"
        style={
          Object {
            "transform": "translateX(0px)",
            "transition": "transform .4s ease-in-out",
          }
        }
      >
        <Image
          alt="image 1"
          key="image 1"
          onError={[Function]}
          onLoad={[Function]}
          src="images/slide-1.jpg"
        />
        <Image
          alt="image 2"
          key="image 2"
          onError={[Function]}
          onLoad={[Function]}
          src="images/slide-2.jpg"
        />
        <Image
          alt="image 3"
          key="image 3"
          onError={[Function]}
          onLoad={[Function]}
          src="images/slide-3.jpg"
        />
      </div>
    </Swiper>
  </div>
  <Icon
    color="lightgray"
    onClick={[Function]}
    style={
      Object {
        "cursor": "pointer",
        "maxWidth": 50,
        "minWidth": 25,
        "width": "7vw",
      }
    }
    value="arrowRight"
    width={60}
  />
</section>
`;
